pkg/config/reconciler.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package config
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// Reconciler is responsible for reconciling Kapp-controllers config.
+type Reconciler struct {
+	coreClient kubernetes.Interface
+	log        logr.Logger
+}
+
+// NewReconciler constructs new Reconciler.
+func NewReconciler(coreClient kubernetes.Interface, log logr.Logger) *Reconciler {
+	return &Reconciler{coreClient, log}
+}
+
+var _ reconcile.Reconciler = &Reconciler{}
+
+// AttachWatches configures watches needed for reconciler to reconcile the kapp-controller Config.
+func (r *Reconciler) AttachWatches(controller controller.Controller) error {
```

Comment:
instead of loading kubeConfig here again, may be we should just pass namespace into this method from outside?

Reply:
Hmm, the namespace isn't available outside the method. `Options` struct contains the optional ns to watch, but this could be `""` which would be weird and also possible different from `GetConfig()` which uses the kubeconfig on disk to get the NS.

I guess to guarantee it's the same, I was opting for loading it again in `AttachWatches`. I'm hesitant to change the `GetConfig()` signature to return the ns it's checking in as it's used in a few places.

Reply:
how about using https://github.com/vmware-tanzu/carvel-kapp-controller/blob/2833f23be56ee9c028d5638bca94bf06f5aabec3/pkg/apiserver/apiserver.go#L44. seems like it's set in kc deployment "dynamically".